packages/gas-tank/src/PaymasterSystem.sol

```diff
@@ -0,0 +1,226 @@
+
+import { System } from "@latticexyz/world/src/System.sol";
+import { Balances as WorldBalances } from "@latticexyz/world/src/codegen/index.sol";
+import { ROOT_NAMESPACE_ID } from "@latticexyz/world/src/constants.sol";
+import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";
+import { RESOURCE_SYSTEM } from "@latticexyz/world/src/worldResourceTypes.sol";
+import { Unstable_CallWithSignatureSystem } from "@latticexyz/world-modules/src/modules/callwithsignature/Unstable_CallWithSignatureModule.sol";
+import { validateCallWithSignature } from "@latticexyz/world-modules/src/modules/callwithsignature/validateCallWithSignature.sol";
+import { EntryPoint } from "./codegen/tables/EntryPoint.sol";
+import { UserBalances } from "./codegen/tables/UserBalances.sol";
+import { Spender } from "./codegen/tables/Spender.sol";
+import { IAllowance } from "./IAllowance.sol";
+
+uint256 constant POST_OP_OVERHEAD = 40000;
+ResourceId constant PAYMASTER_SYSTEM_ID = ResourceId.wrap(
+  bytes32(abi.encodePacked(RESOURCE_SYSTEM, bytes14(""), bytes16("PaymasterSystem")))
+);
+
+/**
+ * @title GasTank PaymasterSystem
+ * @author MUD (https://mud.dev) by Lattice (https://lattice.xyz)
+ * @dev This system implements an ERC4337 paymaster that allows users to share gas balance between different spenders.
+ */
+contract PaymasterSystem is System, IPaymaster, IAllowance {
+  using UserOperationLib for PackedUserOperation;
+  using SimpleAccountUserOperationLib for PackedUserOperation;
+
+  /// @inheritdoc IPaymaster
+  function validatePaymasterUserOp(
+    PackedUserOperation calldata userOp,
+    bytes32 userOpHash,
+    uint256 maxCost
+  ) external override returns (bytes memory context, uint256 validationData) {
+    _requireFromEntryPoint();
+    return _validatePaymasterUserOp(userOp, userOpHash, maxCost);
+  }
+
+  /**
+   * Validate a user operation.
+   * @param userOp     - The user operation.
+   * @param userOpHash - The hash of the user operation.
+   * @param maxCost    - The maximum cost of the user operation.
+   */
+  function _validatePaymasterUserOp(
+    PackedUserOperation calldata userOp,
+    bytes32 userOpHash,
+    uint256 maxCost
+  ) internal virtual returns (bytes memory context, uint256 validationData) {
+    (userOpHash); // unused parameter
+
+    // Require the sender to be a registered spender of a user account
+    address userAccount = Spender.getUserAccount(userOp.getSender());
+    if (userAccount == address(0)) {
+      // Allow calls to `callWithSignature` on this paymaster before a spender is registered
+      userAccount = _recoverCallWithSignature(userOp);
+      if (userAccount == address(0)) {
+        revert("Account is not registered as spender for any user");
+      }
+    }
+
+    // Require the user account to have sufficient balance
+    uint256 balance = UserBalances.get(userAccount);
+    if (maxCost > balance) {
+      revert("Insufficient user balance");
```

Comment:
Should we have custom errors?

Reply:
the reason i went with string errors here is that unfortunately these errors are not surfaced to users anyway (the entry point contract doesn't forward them) and string errors make it easier to debug what's going wrong in the tx trace if one doesn't have the ABI handy